solutions/usaco-624.mdx

```diff
@@ -2,21 +2,175 @@
 id: usaco-624
 source: USACO Platinum 2016 February
 title: Load Balancing
-author: Benjamin Qi
+author: Benjamin Qi, Dustin Miao
 ---
 
+# Solution 1
+
 ## Explanation
 
 The
 [official solution](http://www.usaco.org/current/data/sol_balancing_platinum_feb16.html)
-uses a segment tree, which is introduced in Gold. However, we can just binary
-search on the answer.
+uses a technique known as walking on segment tree. It is perhaps easier to think
+of this as finding the minimum of a function.
+
+Let us partition the points into two sets by x-coordinate. Denote these two sets
+as $L$ and $R$. Because there are up to $\mathcal{O}(n)$ ways to do this, it
+implies that we need to be able partition the points by y-coordinate in
+$\mathcal{O}(1)$ or $\mathcal{O}(\log n)$.
+
+For now, consider just $L$. If we divide it with some line $y = a$, then we can
+create a function
+$f_L(a) = \max(\text{points above } a, \text{points below } a)$. To efficiently
+determine this quantity, we will use a range query data structure like a segment
+tree built on the y-coordinates. Define $f_R(a)$ in the same way.
+
+Still considering just a single set, a key observation is that $f_L$ is a
+_parabolic_ function, in that it is decreasing, attains a minimum value, and
+then increases. Note that parabolic is in name only; this does not imply that
+there is a way to interpolate the points to satisfy an equation $ax^2 + bx + c$.
+We can use ternary search to find the lowest value. Likewise, $f_R$ also
+satisfies this property.
+
+Define a function $f(a) = \max(f_L(a), f_R(a))$. $f$ is also parabolic in nature
+(it helps to draw this out on paper). Thus, we can also use ternary search to
+find the minimum point for $f$.
+
+**Time Complexity:** $\mathcal{O}(N\log N)$
 
 ## Implementation
 
+<LanguageSection>
+
+<CPPSection>
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+#define X first
+#define Y second
+
+const int MAXN = 1e5;
+
+int N;
+pair<int, int> A[MAXN];
+
+struct SegTree {
```

Comment:
this code doesn't even compile? confused

Reply:
oop, it should work now